tilequeue/rawr.py

```diff
@@ -0,0 +1,254 @@
+    assert isinstance(coord, Coordinate)
+    return Tile(coord.zoom, coord.column, coord.row)
+
+
+def convert_to_coord_ints(coords):
+    for coord in coords:
+        coord_int = coord_marshall_int(coord)
+        yield coord_int
+
+
+class RawrToiIntersector(object):
+
+    """
+    Explode and intersect coordinates with the toi
+
+    Prior to enqueueing the coordinates that have had their rawr tile
+    generated, the list should get intersected with the toi.
+    """
+
+    def __init__(self, s3_client, bucket, key):
+        self.s3_client = s3_client
+        self.bucket = bucket
+        self.key = key
+        # state to avoid pulling down the whole list every time
+        self.prev_toi = None
+        self.etag = None
+
+    def tiles_of_interest(self):
+        """conditionally get the toi from s3"""
+        get_options = dict(
+            Bucket=self.bucket,
+            Key=self.key,
+        )
+        if self.etag:
+            get_options['IfNoneMatch'] = self.etag
+        try:
+            resp = self.s3_client.get_object(**get_options)
+        except Exception as e:
+            # boto3 client treats 304 responses as exceptions
+            if isinstance(e, ClientError):
+                resp = getattr(e, 'response', None)
+                assert resp
+            else:
+                raise e
+        status_code = resp['ResponseMetadata']['HTTPStatusCode']
+        if status_code == 304:
+            assert self.prev_toi
+            toi = self.prev_toi
+        elif status_code == 200:
+            body = resp['Body']
+            try:
+                gzip_payload = body.read()
+            finally:
+                try:
+                    body.close()
+                except:
+                    pass
+            gzip_file_obj = StringIO(gzip_payload)
+            toi = load_set_from_gzipped_fp(gzip_file_obj)
+            self.prev_toi = toi
+            self.etag = resp['ETag']
+        else:
+            assert 0, 'Unknown status code from toi get: %s' % status_code
+
+        return toi
+
+    def __call__(self, coords):
+        toi = self.tiles_of_interest()
+        coord_ints = convert_to_coord_ints(coords)
+        intersected_coord_ints, _ = explode_and_intersect(coord_ints, toi)
+        for coord_int in intersected_coord_ints:
+            coord = coord_unmarshall_int(coord_int)
+            yield coord
+
+
+class RawrTileGenerationPipeline(object):
+
+    """Entry point for rawr process command"""
+
+    def __init__(
+            self, rawr_queue, msg_marshaller, group_by_zoom, rawr_gen,
+            queue_writer, rawr_toi_intersector):
+        self.rawr_queue = rawr_queue
+        self.msg_marshaller = msg_marshaller
+        self.group_by_zoom = group_by_zoom
+        self.rawr_gen = rawr_gen
+        self.queue_writer = queue_writer
+        self.rawr_toi_intersector = rawr_toi_intersector
+
+    def __call__(self):
+        # TODO - should probably provide a way to stop iteration
```

Comment:
I agree :wink:

Reply:
Want this added as part of this pr? Would a simple boolean, eg `stopped` or `done`, on the instance suffice?

Reply:
Actually, I take that back.

In `tilequeue_process`, we never stop - just trap a bunch of signals to shut down the threads. In this case, since there aren't any threads, we don't need to do that and a `^C` will suffice to shut it down cleanly enough.

Perhaps we don't need a way to stop iteration? Would we run this as a separate thread / process under the main `tilequeue_process`? Or as a separate service managed by runit?

Reply:
> Perhaps we don't need a way to stop iteration? Would we run this as a separate thread / process under the main tilequeue_process? Or as a separate service managed by runit?

Separate service managed by runit. So yea, we technically don't need to stop iteration, because it's single threaded. Maybe we add some kind of stop mechanism if we add threads?

Reply:
:+1: Sounds good to me.